62 - PHP7 and Nested collections

mpenick opened the pull request.

Commit:
Added nested collection support

Commit:
Fixed existing tests for nested collection changes

Commit:
Fixes and additional unit tests for nested collections

Commit:
Updated PHP docs skeleton for nested collections

Commit:
Reverted changes to composer.lock

Commit:
Changed varint/decimal hash function to use mpz_getlimbn()

Commit:
PHP7 conversion

Commit:
PHP7 conversion fixes

Commit:
PHP5 fixes for PHP7 conversion

mikefero commented on ext/config.m4:
```diff
@@ -89,6 +89,7 @@ if test "$PHP_CASSANDRA" != "no"; then
     src/Cassandra/Set.c \
     src/Cassandra/Map.c \
     src/Cassandra/Collection.c \
+    src/Cassandra/Value.c \
```
This should be added to `ext/config.w32` for Windows builds too.

mikefero commented on ext/doc/Cassandra.php:
```diff
@@ -322,7 +322,12 @@
+    /**
+     * Version of the cpp-driver the extension is compiled against.
+     */
+    const CPP_DRIVER_VERSION = '2.2.2';
```
This might not necessarily be true in the event a different version of the cpp-driver is used. Can we use the [defines in `cassandra.h`](https://github.com/datastax/cpp-driver/blob/master/include/cassandra.h#L53-L56) instead to be dynamic?

mikefero replied:
Nevermind ... This looks like it is just for the documentation stubs and not what is [built into the library](https://github.com/datastax/php-driver/pull/58/files#diff-d3776f8e62027b140044da1533b29c4fR82)

mikefero commented on ext/php_cassandra.c:
```diff
@@ -388,85 +391,16 @@ PHP_RINIT_FUNCTION(cassandra)
-    zval_ptr_dtor(&CASSANDRA_G(type_uuid));
-    CASSANDRA_G(type_uuid) = NULL;
-  }
+#define TYPE_CODE(m) type_ ## m
```
Very nice :exclamation:

avalanche123 commented on ext/doc/Cassandra/Bigint.php:
```diff
@@ -45,6 +45,13 @@ public static function max() {}
+     *
+     * @return Type
+     */
+    public function type() {}
```
missing `implements Value` ?

avalanche123 commented on ext/doc/Cassandra/Decimal.php:
```diff
@@ -41,6 +41,13 @@
+     *
+     * @return Type
+     */
+    public function type() {}
```
missing `implements Value` ?

avalanche123 replied:
ignore, it is `Numeric`

avalanche123 commented on ext/doc/Cassandra/Float.php:
```diff
@@ -45,6 +45,13 @@ public static function max() {}
+     *
+     * @return Type
+     */
+    public function type() {}
```
missing `implements Value` ?

avalanche123 replied:
ignore, it is `Numeric`

avalanche123 commented on ext/doc/Cassandra/Numeric.php:
```diff
@@ -27,7 +27,7 @@
  * @see Varint
  */
-interface Numeric
+interface Numeric implements Value
```
should be `extends` here, since `Numeric` is also an interface, then disregard previous comments

avalanche123 commented on ext/doc/Cassandra/UuidInterface.php:
```diff
@@ -21,7 +21,7 @@
  * A PHP representation of the CQL `timeuuid` datatype
  */
-interface UuidInterface
+interface UuidInterface implements Value
```
also `extends`

avalanche123 commented:
The changes look good minus two things I've pointed out.
As a design suggestion, I'd prefer to keep `Numeric` and `UuidInterface` interfaces unaware of the `Value` interface and implement all interfaces in concrete classes themselves because those interfaces are unrelated.

mikefero commented on ext/src/Cassandra/Timeuuid.c:
```diff
@@ -147,12 +157,12 @@ php_cassandra_timeuuid_properties(zval *object TSRMLS_DC)
 
   char* string = emalloc((CASS_UUID_STRING_LENGTH) * sizeof(char));
```
Is `string` or `self->uuid` freed when calling `efree(self)` in [`php_cassandra_timeuuid_free`](https://github.com/datastax/php-driver/pull/58/files#diff-48d0a9b7bb2a2633ad51f5b3e165e6ebR207)? Just checking in case we have a memory leak.

mikefero commented on ext/util/collections.c:
```diff
@@ -15,160 +17,507 @@
+  switch (type->type) {
   case CASS_VALUE_TYPE_VARCHAR:
   case CASS_VALUE_TYPE_TEXT:
+  case CASS_VALUE_TYPE_ASCII:
```
Good catch.

mikefero commented on ext/src/Cassandra/Collection.c:
```diff
@@ -42,18 +46,20 @@ php_cassandra_collection_find(cassandra_collection* collection, zval* object, lo
+  type = (cassandra_type_collection*) zend_object_store_get_object(collection->type TSRMLS_CC);
 
-  if (!php_cassandra_validate_object(object, collection->type TSRMLS_CC))
+  if (!php_cassandra_validate_object(object, type->value_type TSRMLS_CC))
```
Since the return is on the next line how do you feel about bracing or inlining the return?

mikefero commented on ext/src/Cassandra/Collection.c:
```diff
@@ -91,36 +97,53 @@ php_cassandra_collection_populate(cassandra_collection* collection, zval* array)
+  zval* value_type;
 
-  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &type, &type_len) == FAILURE) {
+  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z", &value_type) == FAILURE)
```
Same here.

mikefero commented on ext/src/Cassandra/Map.c:
```diff
@@ -28,169 +24,227 @@ php_cassandra_map_set(cassandra_map* map, zval* zkey, zval* zvalue TSRMLS_DC)
+  zval* value_type = NULL;
 
-  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "ss", &key_type, &key_type_len, &value_type, &value_type_len) == FAILURE) {
+  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "zz", &key_type, &value_type) == FAILURE)
```
and here.

mikefero commented on ext/src/Cassandra/Set.c:
```diff
@@ -7,130 +9,162 @@ zend_class_entry *cassandra_set_ce = NULL;
+  zval* value_type;
 
-  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "s", &type, &type_len) == FAILURE) {
+  if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "z", &value_type) == FAILURE)
```
and here.

mikefero commented:
LGTM (pending #59 merge); the updated memory management and code consistency (e,g, using self and type consistently) is an added bonus.

There are a bunch more `if` blocks without bracing or inlined; I was just trying to go for consistency here as that appears to be the predominant style used.

Commit:
Nested collection fixes for PHP7

Commit:
Fixed scalar types memory leak

Commit:
Merged PHP7 changes into nested collections

Commit:
Removed redundant collection subtype methods

Commit:
Zend value dtor fix for PHP7

Commit:
Fixing 32-bit builds

Commit:
Adding PHP v7.0 build support for Windows and fixing Windows build issues

Commit:
Fixed collection comparison for PHP5

Commit:
Build fix for cpp-driver versions < 2.2.0

Commit:
More build fixes for cpp-driver versions < 2.2.0

jippi commented:
Nice with PHP7 support - will this also add support for C\* 3.x ?

emulienfou commented:
Great job, just notice 2 problems with the **install.sh** script:
1. Got `configure: error: Unable to load libcassandra`, due to this line: `rm -f build/lib/libcassandra.{dylib,so}`
2. Got `PHP Warning:  PHP Startup: Unable to load dynamic library '/usr/lib64/php/modules/cassandra.so' - libcassandra.so.2: cannot open shared object file: No such file or directory in Unknown on line 0` error, due to this line: `rm -Rf /tmp/php-driver-installation`. For this error I needed to copy **libcassandra.so.2** from `/tmp/php-driver-installation/lib` folder into `/usr/lib/libcassandra.so.2`

I think there is a problem with the install script. Not expert with that however !!!

Commit:
Updating TravisCI build to use container based infrastructure
 - Removing the use of the install.sh script
 - Using only supported PHP versions
  - http://php.net/supported-versions.php
 - Adding caching for libuv and C/C++ driver versions

mpenick commented:
@avalanche123 Thanks for reviewing. Good idea!

To be clear you're advocating this?

``` php
interface Numeric {
    public function value();
}

interface Value {
    public function type();
}

class Bigint implements Value, Numeric {
    public function type() {
    }
    public function value() {
    }
}
```

mpenick commented:
@jippi When using cpp-driver > 2.2.2 you should be able to to connect to connect to and use C\* 3.0 but you won't be able to use some C\* 2.2/3.0 features (e.g. tinyint/smallint types, date/time types, UNSET, new schema metadata).

jippi commented:
@mpenick okay, so 1.0.1 doesn't support 2.2.2 ? its pinned to 2.2.1 in the submodule from what I can see :)

avalanche123 commented:
@mpenick indeed, your understanding is correct

Commit:
Removed 'Value' interface from 'Numeric' and 'UuidInterface'

Commit:
Require cpp-driver 2.2.2+
